app/controllers/predicates/AuthorisedAction.scala

```diff
@@ -101,36 +104,57 @@ class AuthorisedAction @Inject() ()(implicit
         unauthorized
     }
 
-  private[predicates] def agentAuthentication[A](block: User[A] => Future[Result], mtdItId: String)(implicit
-      request: Request[A],
-      hc: HeaderCarrier): Future[Result] = {
+  private[predicates] def agentAuthPredicate(mtdId: String): Predicate =
+    Enrolment(EnrolmentKeys.Individual)
+      .withIdentifier(EnrolmentIdentifiers.individualId, mtdId)
+      .withDelegatedAuthRule(DelegatedAuthRules.agentDelegatedAuthRule)
 
-    lazy val agentDelegatedAuthRuleKey = "mtd-it-auth"
-
-    lazy val agentAuthPredicate: String => Enrolment = identifierId =>
-      Enrolment(EnrolmentKeys.Individual)
-        .withIdentifier(EnrolmentIdentifiers.individualId, identifierId)
-        .withDelegatedAuthRule(agentDelegatedAuthRuleKey)
+  private[predicates] def secondaryAgentPredicate(mtdId: String): Predicate =
+    Enrolment(EnrolmentKeys.SupportingAgent)
+      .withIdentifier(EnrolmentIdentifiers.individualId, mtdId)
+      .withDelegatedAuthRule(DelegatedAuthRules.supportingAgentDelegatedAuthRule)
 
+  private[predicates] def agentAuthentication[A](block: User[A] => Future[Result], mtdItId: String)(implicit
+      request: Request[A],
+      hc: HeaderCarrier): Future[Result] =
     authorised(agentAuthPredicate(mtdItId))
-      .retrieve(allEnrolments) { enrolments =>
-        enrolmentGetIdentifierValue(EnrolmentKeys.Agent, EnrolmentIdentifiers.agentReference, enrolments) match {
-          case Some(arn) =>
-            block(User(mtdItId, Some(arn)))
-          case None =>
-            logger.info("[AuthorisedAction][agentAuthentication] Agent with no HMRC-AS-AGENT enrolment.")
+      .retrieve(allEnrolments) {
+        populateAgent(block, mtdItId, _, isSupportingAgent = false)
+      }
+      .recoverWith(agentRecovery(block, mtdItId))
+
+  private def agentRecovery[A](block: User[A] => Future[Result], mtdItId: String)(implicit
+      request: Request[A],
+      hc: HeaderCarrier): PartialFunction[Throwable, Future[Result]] = {
+    case _: NoActiveSession =>
+      logger.info(s"[AuthorisedAction][agentAuthentication] - No active session.")
+      unauthorized
+    case _: AuthorisationException =>
+      if (appConfig.emaSupportingAgentsEnabled) {
+        authorised(secondaryAgentPredicate(mtdItId))
+          .retrieve(allEnrolments) {
+            populateAgent(block, mtdItId, _, isSupportingAgent = true)
+          }
+          .recoverWith { case _ =>
+            logger.info(s"[AuthorisedAction][agentAuthentication] - Agent does not have secondary delegated authority for Client.")
```

Comment:
please refer the pr for tailor-return-frontend, we need to add the exception handing for Auth as well as downstream error, instead of _